### PR TITLE
Load and run visual plugin (system) on GUI side

### PR DESCRIPTION
**Pull request by iche033**
# 🎉 New feature

Related PR: https://github.com/ignitionrobotics/ign-rendering/pull/533 - Needed for running the `shader_param.sdf` demo world added in this PR, and also for `shader_param_system` integration test to pass.
* Update: https://github.com/ignitionrobotics/ign-rendering/pull/533 is merged and ign-rendering 6.2.0 is released.  `shader_param_system` integration test now passes

## Summary

This PR allows visual plugins to be loaded on the gui side. 

Changes include:
* adding a `VisualPlugin` component to store the `<plugin>` sdf element
  * this component is sent from server to gui
* adding a `VisualPlugin` gui event for `GzSceneManager` to notify `GuiRunner` that a new visual plugin is detected
* extending `GuiRunner` to load and run systems (visual plugin)
  * this can probably be refactored later to avoid duplicate code with `SimulationRunner`
* adding a `SceneUpdate` event that systems can connect to on both the server and gui side for rendering
* adding a `ShaderParam` visual plugin (ported from gazebo-classic) to test the whole pipeline

## Test it

Run the `shader_param.sdf` world - this world contains a sphere model with the new `ShaderParam` system that modifies its appearance using custom shaders. Requires https://github.com/ignitionrobotics/ign-rendering/pull/533

```
ign gazebo -v 4  -r shader_param.sdf
```

shader_param

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [x] Added example and/or tutorial
- [x] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

File: include/ignition/gazebo/components/Visual.hh [hunk 1]
```diff
@@ -17,6 +17,9 @@
 #ifndef IGNITION_GAZEBO_COMPONENTS_VISUAL_HH_
 #define IGNITION_GAZEBO_COMPONENTS_VISUAL_HH_
 
+#include <string>
+#include <sdf/Element.hh>
+#include <sdf/parser.hh>
```
Comment (ahcorde) [inline include/ignition/gazebo/components/Visual.hh:22]: alphabetize

File: include/ignition/gazebo/components/Visual.hh [hunk 2]
```diff
@@ -27,11 +30,60 @@ namespace gazebo
 {
 // Inline bracket to help doxygen filtering.
 inline namespace IGNITION_GAZEBO_VERSION_NAMESPACE {
+namespace serializers
+{
+  class SdfElementSerializer
+  {
+    /// \brief Serialization for `sdf::Model`.
+    /// \param[in] _out Output stream.
+    /// \param[in] _time Model to stream
```
Comment (ahcorde) [inline include/ignition/gazebo/components/Visual.hh:39]:
```suggestion
    /// \param[in] _elem Model to stream
```

File: include/ignition/gazebo/components/Visual.hh [hunk 3]
```diff
@@ -27,11 +30,60 @@ namespace gazebo
...
+           << "</sdf>";
+      return _out;
+    }
+
+    /// \brief Deserialization for `sdf::Element`.
+    /// \param[in] _in Input stream.
+    /// \param[out] _model Model to populate
+    /// \return The stream.
+    public: static std::istream &Deserialize(std::istream &_in,
+                sdf::ElementPtr &_elem)
```
Comment (ahcorde) [inline include/ignition/gazebo/components/Visual.hh:56]:
```suggestion
                sdf::ElementPtr &_model)
```

File: include/ignition/gazebo/gui/GuiEvents.hh [hunk 4]
```diff
@@ -25,6 +25,7 @@
 #include <string>
 #include <utility>
 #include <vector>
+#include <sdf/Element.hh>
```
Comment (ahcorde) [inline include/ignition/gazebo/gui/GuiEvents.hh:28]: alphabetize

File: include/ignition/gazebo/gui/GuiEvents.hh [hunk 5]
```diff
@@ -211,6 +212,28 @@ namespace events
     IGN_UTILS_IMPL_PTR(dataPtr)
   };
 
+  /// \brief Event that notifies a visual plugin is to be loaded
+  class IGNITION_GAZEBO_GUI_VISIBLE VisualPlugin: public QEvent
+  {
+    /// \brief Constructor
+    /// \param[in] _entity Entity added
+    /// \param[in] _type Entity type
+    /// \param[in] _parent Parent entity.
```
Comment (ahcorde) [inline include/ignition/gazebo/gui/GuiEvents.hh:221]:
```suggestion
```

File: src/gui/plugins/scene_manager/GzSceneManager.cc [hunk 6]
```diff
@@ -14,10 +14,15 @@
  * limitations under the License.
  *
 */
+
+#include <map>
 #include <set>
 
+#include "../../GuiRunner.hh"
```
Comment (ahcorde) [inline src/gui/plugins/scene_manager/GzSceneManager.cc:21]: is it fine to add relative paths ?

File: src/systems/shader_param/ShaderParam.cc [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+class ignition::gazebo::systems::ShaderParamPrivate
+{
+  /// \brief Data structure for storing shader param info
+  public: class ShaderParamValue
+  {
+    /// \brief shader type: vertex or fragment
+    public: std::string shader;
+
+    /// \brief variable type: int, float, float_array, int_array
+    /// \todo support samplers
```
Comment (ahcorde) [inline src/systems/shader_param/ShaderParam.cc:55]:
```suggestion
    /// TODO(anyone): support samplers
```

File: test/integration/shader_param_system.cc [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+/////////////////////////////////////////////////
+// The test checks camera image data to verify that the sphere is using
+// custom material shaders
+TEST_F(ShaderParamTest, IGN_UTILS_TEST_DISABLED_ON_MAC(ShaderParamBox))
+{
+  // Start server
+  ServerConfig serverConfig;
+  const auto sdfFile = std::string(PROJECT_SOURCE_PATH) +
+    "/examples/worlds/shader_param.sdf";
```
Comment (ahcorde) [inline test/integration/shader_param_system.cc:68]: ignition::common::joinPaths

File: test/integration/shader_param_system.cc [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+    sleep++;
+  }
+  EXPECT_LT(sleep, maxSleep);
+  ASSERT_NE(imageBuffer, nullptr);
+
+  // shaders set the sphere color to red
+  unsigned int height = 320;
+  unsigned int width = 240;
+
+  int mid = (height/2 * width * 3u) + (width/2 - 1) * 3u;
```
Comment (ahcorde) [inline test/integration/shader_param_system.cc:98]:
```suggestion
  int mid = (height / 2 * width * 3u) + (width / 2 - 1) * 3u;
```

File: test/plugins/TestVisualSystem.hh [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+        {
+          igndbg << "Destroying TestVisualSystem" << std::endl;
+        }
+
+  private: bool Service(msgs::StringMsg &_msg)
+           {
+             igndbg << "TestVisualSystem service called" << std::endl;
+             _msg.set_data("TestVisualSystem");
+             return true;
+           }
```
Comment (ahcorde) [inline test/plugins/TestVisualSystem.hh:59]:
```suggestion
        {
           igndbg << "TestVisualSystem service called" << std::endl;
           _msg.set_data("TestVisualSystem");
           return true;
        }
```

**Review by ahcorde — CHANGES_REQUESTED** [review]

**Comment by chapulina** [comment]
I believe this closes https://github.com/ignitionrobotics/ign-gazebo/issues/657

File: include/ignition/gazebo/components/Visual.hh (hunk 2 above)
Comment (iche033) [reply include/ignition/gazebo/components/Visual.hh]: fixed. 9f4a753763f64918d2441b56f16cb8be9040c699

File: include/ignition/gazebo/components/Visual.hh (hunk 3 above)
Comment (iche033) [reply include/ignition/gazebo/components/Visual.hh:58]: fixed. 9f4a753763f64918d2441b56f16cb8be9040c699

File: include/ignition/gazebo/gui/GuiEvents.hh (hunk 4 above)
Comment (iche033) [reply include/ignition/gazebo/gui/GuiEvents.hh]: added line to separate the system headers and ignition / sdf headers and arranged them in alphabetical order. 9f4a753763f64918d2441b56f16cb8be9040c699

File: include/ignition/gazebo/gui/GuiEvents.hh (hunk 5 above)
Comment (iche033) [reply include/ignition/gazebo/gui/GuiEvents.hh]: fixed. 9f4a753763f64918d2441b56f16cb8be9040c699

File: src/gui/plugins/scene_manager/GzSceneManager.cc (hunk 6 above)
Comment (iche033) [reply src/gui/plugins/scene_manager/GzSceneManager.cc:21]: I think so. It's was done here as well: https://github.com/ignitionrobotics/ign-gazebo/blob/ign-gazebo6/src/gui/plugins/modules/EntityContextMenu.cc#L18

File: test/integration/shader_param_system.cc (hunk 9 above)
Comment (iche033) [reply test/integration/shader_param_system.cc]: 9f4a753763f64918d2441b56f16cb8be9040c699

File: test/integration/shader_param_system.cc (hunk 8 above)
Comment (iche033) [reply test/integration/shader_param_system.cc]: 9f4a753763f64918d2441b56f16cb8be9040c699

File: src/systems/shader_param/ShaderParam.cc (hunk 7 above)
Comment (iche033) [reply src/systems/shader_param/ShaderParam.cc]: done. 9f4a753763f64918d2441b56f16cb8be9040c699

File: test/plugins/TestVisualSystem.hh (hunk 10 above)
Comment (iche033) [reply test/plugins/TestVisualSystem.hh]: 9f4a753763f64918d2441b56f16cb8be9040c699

File: include/ignition/gazebo/components/Visual.hh (hunk 1 above)
Comment (iche033) [reply include/ignition/gazebo/components/Visual.hh]: hmm I think this means alphabetize and also taking into account case-sensitivity? If so, done in 9f4a753763f64918d2441b56f16cb8be9040c699

File: src/gui/GuiRunner.cc [hunk 11]
```diff
@@ -264,12 +309,102 @@ void GuiRunner::OnStateQt(const msgs::SerializedStepMap &_msg)
...
+      {
+        SystemPluginPtr sys = system.value();
+        this->dataPtr->systems.push_back(sys);
+        this->dataPtr->systemsPreupdate.push_back(
+            sys->QueryInterface<ISystemPreUpdate>());
+        this->dataPtr->systemsUpdate.push_back(
+            sys->QueryInterface<ISystemUpdate>());
+        this->dataPtr->systemsPostupdate.push_back(
+            sys->QueryInterface<ISystemPostUpdate>());
+
```
Comment (ahcorde) [inline src/gui/GuiRunner.cc:355]:
```suggestion
```

File: src/gui/plugins/scene_manager/GzSceneManager.cc [hunk 12]
```diff
@@ -58,6 +64,8 @@ inline namespace IGNITION_GAZEBO_VERSION_NAMESPACE {
     /// \brief Mutex to protect gui event and system upate call race conditions
     /// for newEntities and removedEntities
     public: std::mutex newRemovedEntityMutex;
+
+    public: bool initializedVisualPlugins = false;
```
Comment (ahcorde) [inline src/gui/plugins/scene_manager/GzSceneManager.cc:68]: doc ?

File: src/systems/shader_param/ShaderParam.cc [hunk 13]
```diff
@@ -0,0 +1,396 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "ShaderParam.hh"
+
+#include <list>
```
Comment (ahcorde) [inline src/systems/shader_param/ShaderParam.cc:20]: alphabetize

**Review by ahcorde — APPROVED** [review]
LGTM, just some minor style issues

File: src/systems/shader_param/ShaderParam.cc (hunk 13 above)
Comment (iche033) [reply src/systems/shader_param/ShaderParam.cc]: 0dead0578ba3f26cf97f8162e55a0dbe1112484c

File: src/gui/plugins/scene_manager/GzSceneManager.cc (hunk 12 above)
Comment (iche033) [reply src/gui/plugins/scene_manager/GzSceneManager.cc:69]: 0dead0578ba3f26cf97f8162e55a0dbe1112484c

**Comment by srmainwaring** [comment]
@iche033 - just checked and that this works for Metal with https://github.com/ignitionrobotics/ign-rendering/pull/554 🙂 

visual_plugin

Here are the metal shaders:

[deform_fs.metal.zip](https://github.com/ignitionrobotics/ign-gazebo/files/8019422/deform_fs.metal.zip)
[deform_vs.metal.zip](https://github.com/ignitionrobotics/ign-gazebo/files/8019423/deform_vs.metal.zip)

**Comment by iche033** [comment]
oh nice, looks like you forward ported this to Garden? I can update the shader param system to read metal shaders, e.g. something like:

```
<shader launguage="glsl">
  <vertex>path_to_vertex_shader.glsl</vertex>
  <fragment>path_to_fragment_shader.glsl</fragment>
</shader> 
<shader launguage="metal">
  <vertex>path_to_vertex_shader.metal</vertex>
  <fragment>path_to_fragment_shader.metal</fragment>
</shader> 
```

This just lets user specify multiple shader files and the system will then select the best shader to use depending on the platform. How does that look?

**Comment by iche033** [comment]
Implemented the above idea in the [shader_param_metal](https://github.com/ignitionrobotics/ign-gazebo/commit/17dc4ae5b5193c1fc1d4d26e73f288d2cbbabe5a) branch. I'll create a PR once I can [upload metal shaders to Fuel](https://gitlab.com/ignitionrobotics/web/app/-/merge_requests/148). 

I've also created https://github.com/ignitionrobotics/ign-rendering/pull/558 to accomodate cases like https://github.com/ignitionrobotics/ign-rendering/pull/554 where extra params (`shader_reflection_pair_hint`) are needed for `metal` but not `glsl`

**Comment by srmainwaring** [comment]
> oh nice, looks like you forward ported this to Garden?

Yes, just cherry-picked it in. Most of the Metal support needs Garden now because of various ABI breaking interface changes.

Your proposed changes for multi-shader language support and extra param checks look good!

**Comment by srmainwaring** [comment]
@iche033 Just noticed you've already done all the texture changes in #1310! I should have looked more carefully...

**Comment by osrf-triage** [comment]
This pull request has been mentioned on **Gazebo Community**. There might be relevant details there:

https://community.gazebosim.org/t/new-ignition-releases-2022-03-01-citadel-edifice-fortress/1313/1

**Comment by osrf-triage** [comment]
This pull request has been mentioned on **Gazebo Community**. There might be relevant details there:

https://community.gazebosim.org/t/using-a-custom-hlms-with-ignition-rendering/1315/2

**Pull request merged**